### DISCUSSION TITLE
Another take on integrating MDM with template plugins

### DISCUSSION THREAD

**Discussion by Jawabiscuit** [mdelobelle/metadatamenu#543] [Ideas]
Hi, I'm a new user of MDM and I agree with #307 made a year ago, in that it would be highly valuable to be able to have an interface for our curated metadata fields at file creation time. Most of the pain I'm experiencing recently when trying to maintain consistency has come from the creation context, where asking for input through suggesters and prompts from TP or QA is very repetitive and the modals lack visual/functional cohesiveness. I've thought about alternatives outside of using MDM but there are pitfalls related to the following.

1. Managing metadata with multiple plugins
2. Support for MDM data types being shoe-horned into another plugin where it doesn't belong
3. Some other ad-hoc solution that is fragile to plugin changes

Here's my take on what the experience from a user standpoint could be. The consumer of the API call could be Templater, QuickAdd, or JS. The modal would look similar to the `metadata-menu:open_fields_modal` with the "FileClass Select" option enabled, as well as the ability to multi-select file classes. The user should be free to omit selecting an option for any fields and that be reflected in the output.

```mermaid
sequenceDiagram
    participant U as User
    participant MM as Metadata Menu
    participant TP as Templater
    Note over U: 1. Create a new note
    U->>TP: 
    MM->>MM: fields from settings and file classes
    MM->>MM: default options
    TP->>MM: fileFieldsModal() or<br>fileFields({modal: true}) API
    Note over U: 2. Select type & field values<br>(modal does not insert)
    MM->>U: Fields modal
    U->>MM: Desired values
    Note left of TP: Output is similar to fileFields()<br>with indexedPath undefined 
    MM->>TP: fields
    TP->>TP: other logic
    TP->>TP: format fields
    Note over U: Done
```

Thanks for reading, and thanks for creating such a robust and valuable plugin! I'll happily donate once I'm off unemployment.

**Comment by LeonardoGentile** [comment]
Is there any plan to implement this?
Many of us are in a similar situation where we need to rely on templater or quickadd that are not aware of metadata fields and types. Being able to create new notes with defaults and modals for a specific fileclass from within mdm would be perfect

**Reply by LeonardoGentile** [reply]
As a workaround for the time being, wouldn’t it be possible to let templater interact wit mdm via an api to fetch a specific schema in order to populate a template for a note of a certain class? If yes, how? This would avoid metadata duplication and sync (classes definition and templates definition)

**Reply by Jawabiscuit** [reply]
In short, I think the answer is yes, but IMO the potential solution this issue proposes wouldn't be asking for much more and with significantly better potential UX. Here's my workaround for now. Hopefully it can serve as proof-of-concept or at least evidence there is a gap that could be filled.

❗ Warning: What follows is not intended to be shared and is an abomination. At least the code should be easy to read, with plenty of comments. 😆 ❗ 

I built a wrapper API, "glue code" might be more appropriate which gets called directly when creating a new file, [prompting you to select a fileClass]. All the fileClass metadata is compiled [straight from their file frontmatter] which are gotten from [MDM's settings]. Then the fields are merged similar to how you'd expect MDM to, though I know it's doing it according to how I currently use MDM and it isn't an exact match, which is a point of failure, and I also haven't looked deeply into MDM's code. The complexity then comes when prompting for their input. I believe all these responsibilities are handled more neatly and as intended under the hood of MDMs fields modal. It is also more compact because, as you can see in my current solution there is a lot of [logic to determine which TP suggester to invoke]--and many suggesters can potentially be invoked! That last part I considered using [modal-forms] plugin for but it still would be painful to configure, and introduces more complexity, so I opted to keep the dependencies lean.

My solution is far from complete and I'm not proud of it, but handles the basics, I'm hoping, temporarily. There's a lot of sophistication with nested Object fields it currently can't handle. I'm also providing more options in a very inaccessible way for each fileClass that doesn't exist currently with MDMs file classes which I would make suggestions about to MDM if I knew which direction it was headed. I'd love to look into how it could be implemented directly but I've burned up all my available time to focus on MDM for the time being. I created this issue thinking that it would be a much simpler change for MDM to provide, so more time saved for all parties and a substantial payoff--a win-win.

[prompting you to select a fileClass]: <https://github.com/Jawabiscuit/obsidian-modules/blob/sandbox/metadata.js#L651> "promptForFileClass"
[straight from their file frontmatter]: <https://github.com/Jawabiscuit/obsidian-modules/blob/sandbox/metadata.js#L112> "getFileClassMetadata"
[MDM's settings]: <https://github.com/Jawabiscuit/obsidian-modules/blob/sandbox/metadata.js#L93> "getFileClasses"
[logic to determine which TP suggester to invoke]: <https://github.com/Jawabiscuit/obsidian-modules/blob/sandbox/metadata.js#L339> "getValuesForFields"
[modal-forms]: <https://github.com/danielo515/obsidian-modal-form> "obsidian-modal-forms"

**Reply by Jawabiscuit** [reply]
I made this discussion, not an issue as I said, and idea and I'd really like to keep the ideas flowing! Maybe the maintainer has ideas that are more in line with their vision!?

**Reply by LeonardoGentile** [reply]
@Jawabiscuit man, that's a lot of interesting code I wish was embedded in neat plugin 🤩
[rant-start] it is so painful that there are so many different plugins trying to solve very similar issues, usually related to metadata management and templating, but unfortunately most of them solve just a tiny fraction of the problem and they don't communicate well with each other[rant-stop] 
MDM would be perfect as the _source of truth_ when dealing with metadata so I'm not against separating templating and new file creation into other plugins but they need to easily access and use mdm functionalities. A word from @mdelobelle would indeed be interesting to understand the direction of the development 🙏

**Reply by Jawabiscuit** [reply]
> that's a lot of interesting code

Thanks! I could write a plugin, but then, I'd be married to it and I'd feel a sense of obligation. I'm all for contributing to such a vibrant community, but weighing the rewards of such a relationship is keeping me on the sidelines.

> unfortunately most of them solve just a tiny fraction of the problem and they don't communicate well with each other

Agreed! I started to rant myself, but there's probably better forums for that kind of discussion, Reddit perhaps? 😉
